src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildProject.cs

```diff
@@ -42,6 +42,8 @@ internal partial class MSBuildProject : IProject
 
         public string Id => Context.SolutionInfo.GetProjectId(FileInfo.FullName);
 
+        public IEnumerable<string> AllProjectReferences => GetRoslynProject().AllProjectReferences.Select(projRef => projRef.ProjectId.ToString());
```

Comment:
Seems like there's an optimization we could do here where we're not calling `GetRoslynProject()` every time somebody asks for either this or `.ProjectReferences`. Additionally, the LINQ query in both cases ends up getting executed _every_ time somebody Gets either property.

If either/both of these never change, we should optimize both of them with a private backing field or readonly property set at ctor time.

Reply:
I think this is an optimization we can do for many of the properties and methods here but being carefully making sure that introducing state would not cause any breaks. For example, this is a possibility in cases where the underlying project changes using the `ProjectFile.SaveAsync` method and thus `GetRoslynProject()` returns a different project.

I would be happy to take this as a separate issue, as it might require some calculated optimizations.